File: src/vct/col/ast/ActionBlock.java
package vct.col.ast;

public class ActionBlock extends ASTNode {

  @Override
  public <R,A> R accept_simple(ASTMapping1<R,A> map,A arg){
    return map.map(this,arg);
  }
  
  public final ASTNode history;
  public final ASTNode fraction;
  public final ASTNode process;
  public final ASTNode action;
  public final ASTNode block;
  
  public ActionBlock(ASTNode history,ASTNode fraction,ASTNode process,ASTNode action,ASTNode block){
    this.history=history;
    this.fraction=fraction;
    this.process=process;
    this.action=action;
    this.block=block;
  }
  
  @Override
  public <T> void accept_simple(ASTVisitor<T> visitor){
    try {
      visitor.visit(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }
  
  @Override
  public <T> T accept_simple(ASTMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
      }
      throw t;
    }
  }
 

}

File: vercors/src/main/java/vct/col/ast/ActionBlock.scala
package vct.col.ast

import scala.collection.JavaConverters._
import vct.col.util.VisitorHelper

class ActionBlock(val history:ASTNode, val fraction:ASTNode, val process:ASTNode, val action:ASTNode, val map:Map[String,ASTNode], val block:ASTNode) extends ASTNode with VisitorHelper {
  def mapAsJava() = map.asJava
  
  /** This constructor is added to retain compatibility with Java (by converting the map to a Scala structure) */
  def this(history:ASTNode, fraction:ASTNode, process:ASTNode, action:ASTNode, map:java.util.Map[String,ASTNode], block:ASTNode) = 
    this(history, fraction, process, action, map.asScala.toMap, block)
  
  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))
}